### ISSUE TITLE
Code passing CUDA simulation mode does not work

### ISSUE THREAD

**Issue by mmlanger** [numba/numba#2485]
The problem is if one uses multi dimensional local or shared arrays on cuda devices and writes array expressions.
1D array expressions seem to be working but while the following code works if
NUMBA_ENABLE_CUDASIM is set it won't actually work CUDA mode.

```
import numpy as np

import numba
from numba import cuda

@cuda.jit
def generate_kernel(mem_array):
    i = cuda.grid(1)
    buffer = cuda.local.array((1, 4), numba.float64)

    if i < mem_array.shape[0]:
        buffer[0] = mem_array[0]
        #for k in range(mem_array.shape[1]):
        #    buffer[k] = mem_array[0, k]


n_init_conds = 32 * 4
mem_array = np.zeros((n_init_conds, 4))
mem_array[0] = np.array([0.1, 0.2, 0.3, 0.4])

threadsperblock = 32
blockspergrid = (n_init_conds + (threadsperblock - 1)) // threadsperblock
generate_kernel[threadsperblock, blockspergrid](mem_array)
```

I could solve it be explicitly indexing everything, like in the commented code above, but simulation mode should complain about it.

**Comment by seibert** [comment]
This is a tricky issue, because we created the simulator to deliberately allow non-CUDA code (like starting PDB and printing arbitrary expressions) to run for debugging purposes.  The expectation is that you start from code that compiles for the CUDA target, and then move over to the simulator to investigate.  We should make this more clear in the documentation.

**Comment by esc** [comment]
This sounds like a feature request for additional documentation, cc @gmarkall ?

**Comment by gmarkall** [comment]
> This sounds like a feature request for additional documentation, cc @gmarkall ?

I think it does, yes. Fix proposed in #6244
